Data/src/main/java/org/tribuo/data/csv/CSVLoader.java

```diff
@@ -258,110 +255,65 @@ public CSVLoader(OutputFactory<T> outputFactory) {
-        //
-        // Create the examples.
-        while (itr.hasNext()) {
-            Map<String, String> row = itr.next().getRowData();
-            T label = (responseNames.size() == 1) ?
-                    buildOutput(responseName, row) :
-                    buildMultiOutput(responseNames, row);
-            ArrayExample<T> example = new ArrayExample<>(label);
-            for (Map.Entry<String, String> ent : row.entrySet()) {
-                String columnName = ent.getKey();
-                if (!responseNames.contains(columnName)) {
-                    //
-                    // If it's not a response, it's a feature
-                    double value = Double.parseDouble(ent.getValue());
-                    example.add(columnName, value);
+    public DataSource<T> loadDataSource(URL csvPath, Set<String> responseNames, String[] header) throws IOException {
+        List<String> headers = header == null || header.length == 0 ? Collections.emptyList() : Arrays.asList(header);
+        URI csvURI;
+
+        // Extract headers and convert to URI
+        try (CSVIterator itr = new CSVIterator(csvPath.toURI(), separator, quote)) {
+            List<String> extractedHeaders = itr.getFields();
+            if (extractedHeaders.isEmpty() && headers.isEmpty()) {
+                throw new IllegalArgumentException("Failed to read headers from CSV, and none were supplied.");
+            }
+            if (headers.size() != 0) {
+                if (extractedHeaders.size() != headers.size()) {
+                    throw new IllegalArgumentException("The csv contains " + extractedHeaders.size() + " fields, but only " + headers.size() + " headers were supplied.");
                 }
+            } else {
+                headers = extractedHeaders;
             }
-            dataset.add(example);
+            csvURI = csvPath.toURI();
+        } catch (URISyntaxException e) {
+            throw new FileNotFoundException("Failed to read from URL '" + csvPath + "' as it could not be converted to a URI");
         }
-        return dataset;
-    }
 
-    private static void validateResponseNames(Set<String> responseNames, List<String> headers, String csvPath) throws IllegalArgumentException {
+        // Validate the responseNames
         if (responseNames.isEmpty()) {
             throw new IllegalArgumentException("At least one response name must be specified, but responseNames is empty.");
         }
-        //
-        // Validate that all the expected responses are included in the given header fields
-        Map<String, Boolean> responsesFound = new HashMap<>();
-        for (String response : responseNames) {
-            responsesFound.put(response, false);
-        }
-        for (String header : headers) {
-            if (responseNames.contains(header)) {
-                responsesFound.put(header, true);
+        if (!headers.containsAll(responseNames)) {
```

Comment:
It looks like this is to mimic the existing behavior, but is it desirable from a UX standpoint to have the user enter the response columns in both responses and headers? I feel like the intuition of keeping features and responses separate is strong and that we would want to support that.

Reply:
The headers are used if the CSV file doesn't have headers, and so you need the column ordering.

Reply:
That makes sense.